firedrake/ensemble/ensemble_functionspace.py

```diff
@@ -0,0 +1,242 @@
+    """
+    A mixed function space defined on an :class:`firedrake.Ensemble`.
+    The subcomponents are distributed over the ensemble members, and
+    are specified locally.
+
+    Parameters
+    ----------
+    local_spaces : Collection
+        The list of function spaces on the local ensemble.comm.
+    ensemble : :class:`firedrake.Ensemble`
+        The communicator that the function space is defined over.
+
+    Notes
+    -----
+    Passing a list of dual local_spaces to :class:`firedrake.EnsembleFunctionSpace`
+    will return an instance of :class:`firedrake.EnsembleDualSpace`.
+
+    This class does not carry UFL symbolic information, unlike a
+    :function:`firedrake.FunctionSpace`. UFL expressions can only be
+    defined locally on each ensemble member using a :function:`firedrake.FunctionSpace` from `EnsembleFunctionSpace.local_spaces`.
+
+    See also:
+    - Primal ensemble objects: :class:`firedrake.EnsembleFunctionSpace` and :class:`firedrake.EnsembleFunction`.
+    - Dual ensemble objects: :class:`firedrake.EnsembleDualSpace` and :class:`firedrake.EnsembleCofunction`.
+    """
+    def __init__(self, local_spaces, ensemble):
+        meshes = set(V.mesh() for V in local_spaces)
+        if len(meshes) != 1:
+            raise ValueError(
+                f"{self.__class__.__name__} local_spaces must all be defined on the same mesh.")
+        self._mesh = meshes.pop()
+        self._ensemble = ensemble
+        self._local_spaces = tuple(local_spaces)
+
+        # Internally we'll store everything in a big mixed space. For
+        # EnsembleFunctions/Cofunctions, we'll create (possibly mixed)
+        # subfunctions that view the correct subfunctions of this big space.
+        self._full_local_space = MixedFunctionSpace(self.local_spaces)
+
+        # ensemble._comm is congruent with ensemble.global_comm not ensemble.comm
+        # because obj._comm is used for garbage collection, so it needs to be the
+        # communicator that the ensemble objects are collective over.
+        self._comm = internal_comm(ensemble._comm, self)
+
+    @property
+    def ensemble(self):
+        """The :class:`firedrake.Ensemble` that the function space is defined over
+        """
+        return self._ensemble
+
+    @property
+    def comm(self):
+        """The spatial communicator from the :class:`firedrake.Ensemble` communicator.
+        """
+        return self._ensemble.comm
+
+    @property
+    def ensemble_comm(self):
+        """The ensemble communicator from the :class:`firedrake.Ensemble` communicator.
+        """
+        return self._ensemble.ensemble_comm
+
+    @property
+    def global_comm(self):
+        """The global communicator from the :class:`firedrake.Ensemble` communicator.
+        """
+        return self._ensemble.global_comm
+
+    @property
+    def local_spaces(self):
+        """The :function:`firedrake.FunctionSpace` on the local ensemble.comm.
+        """
+        return self._local_spaces
```

Comment:
Up to you but I personally would probably just set `self.local_spaces` instead of writing a getter.